receiver/mongodbatlasreceiver/internal/mongodb_atlas_client.go

```diff
@@ -626,6 +626,27 @@ func (s *MongoDBAtlasClient) GetClusters(ctx context.Context, groupID string) ([
 	return clusters, nil
 }
 
+// GetAlerts returns the alerts specified for the set projects
+func (s *MongoDBAtlasClient) GetAlerts(ctx context.Context, groupID string) ([]mongodbatlas.Alert, error) {
+	pageNum := 1
```

Comment:
What do we know about the ordering of alerts? Are they sorted by most recent?

If so, can't we just keep track of the most recent time stamp that was previously processed, and only call for pages until we hit that time? If we can do this, we don't need the cache at all - we just need to keep track of a timestamp.

Reply:
If the above is not possible, we should look into whether the [count](https://github.com/mongodb/go-client-mongodb-atlas/blob/d82e1c95c5ba750f0545e408bebfc47e701691eb/mongodbatlas/mongodbatlas.go#L189) option can give us what we need.

If the count represents the total number of alerts _available to request_, rather than the number returned in an individual request, then we calculate how many alerts need to be processed since the last count.

Reply:
> What do we know about the ordering of alerts? Are they sorted by most recent?

My current understanding of them is that they are ordered by most recent however I think what gets somewhat problematic is capturing updates to the alerts as they will appear in order by creation, not update.

i.e. Missing the capture of an acknowledgement or closure of an alert but that remains on page 2 because of the default order by creation. 

My fear is missing an acknowledgement event or an event closure as those would have been captured in the listening mode as well. Happy to explore potential optimizations but I want to ensure that this case is not overlooked.

Reply:
Ok, thanks for clarifying.

It sounds like we might be able to process "new" in a way that does not require much state (just the most recent timestamp of previously processed alerts).

I'm curious what options we have for "updated" alerts. I see these are flagged by timestamp, so we also do not need to cache each event, we just need to check if the timestamp is newer than the most recent we saw in the last collection. 

What's interesting here is that without it being sorted by last updated, it seems we have to page through _all_ alerts, every collection, even if just to check if the timestamp is new. This seems like a poor design of the API, but if that's what we have to work with, then I think we should consider how to put some kind of reasonable limits on this. We probably ought to have a config setting for specifying the max number of events or a time window outside of which we do not search for updates.

In summary, what I'm suggesting is an algorithm that requires no cache, and relies only on a timestamp. Roughly:
- The receiver keeps track of a "most recent alert" time.
- Page through alerts (which goes backwards in time). 
  - When we hit an alert that occurred _before_ the "most recent alert" time, we know we've consumed all the new alerts.
- Continue to page through alerts until we hit the configured cutoff (either number of alerts of end of time window)
  - If an alert has an updated time _after_ the "most recent alert" time, then consumer it. Otherwise, ignore it and move on.
- Update the "most recent time" based on the alerts we consumed this cycle (either new or updated).

Thoughts?

Reply:
Just pushed up a change that should follow this algorithm. 

It stores and uses the `updated` timestamp to decide whether or not to process an alert.

Reply:
> would we still need to store that "most recent time" to keep a line on collector restarts?

Yes, that'd still be necessary to prevent duplicate reporting of alerts. I figure it's much easier and more efficient to manage a single timestamp vs a cache full of hashes though.

Reply:
> If we load a timestamp for one of the alerts and we try to process it again and since it has the same timestamp a time.Before() call would be false and we would try to process the identical alert again.

Ok, so it sounds like we need `<=` instead of just `<`. Surely there is a way to compare timestamps like this.

Reply:
> > If we load a timestamp for one of the alerts and we try to process it again and since it has the same timestamp a time.Before() call would be false and we would try to process the identical alert again.
> 
> Ok, so it sounds like we need `<=` instead of just `<`. Surely there is a way to compare timestamps like this.

But then on subsequent polls we would emit already emitted logs no? I.e the first bullet point?

Reply:
> If we check equality along with it, we drop data between subsequent alerts that have the same updated time as those would have equality in update times.

I don't understand this. Can you elaborate or give an example?

Reply:
While drawing up an example I realized my folly was iterating and updating only one timestamp, what I did to solve it was to keep track of the latest in the payload and separating that comparison.

Here is the solution I came up with:

<details>

<summary> Snippet </summary>

```go
func (a *alertsReceiver) applyFilters(pConf ProjectConfig, alerts []mongodbatlas.Alert) []mongodbatlas.Alert {
	filtered := []mongodbatlas.Alert{}

	var lastRecordedTime = pcommon.Timestamp(0).AsTime()
	if a.record.LastRecordedTime != nil {
		lastRecordedTime = *a.record.LastRecordedTime
	}
	// we need to maintain two timestamps in order to not conflict while iterating
	var latestInPayload time.Time = pcommon.Timestamp(0).AsTime()

	for _, alert := range alerts {
		updatedTime, err := time.Parse(time.RFC3339, alert.Updated)
		if err != nil {
			a.logger.Warn("unable to interpret updated time for alert, expecting a RFC3339 timestamp", zap.String("timestamp", alert.Updated))
			continue
		}

		if updatedTime.Before(lastRecordedTime) || updatedTime.Equal(lastRecordedTime) {
			// already processed if the updated time was before or equal to the last recorded
			continue
		}

		if len(pConf.ExcludeClusters) > 0 {
			var shouldExclude = false
			for _, exclusion := range pConf.ExcludeClusters {
				if alert.ClusterName == exclusion {
					shouldExclude = true
				}
			}
			if shouldExclude {
				continue
			}
		}

		if len(pConf.IncludeClusters) > 0 {
			var shouldInclude = false
			for _, inclusion := range pConf.IncludeClusters {
				if alert.ClusterName == inclusion {
					shouldInclude = true
				}
			}
			if !shouldInclude {
				continue
			}
		}

		filtered = append(filtered, alert)
		if updatedTime.After(latestInPayload) {
			latestInPayload = updatedTime
		}
	}

	if latestInPayload.After(lastRecordedTime) {
		a.record.SetLastRecorded(&latestInPayload)
	}

	return filtered
}
```

<details>

Reply:
I see, you were updating the cutoff time before processing all alerts for a poll interval.